Amazon.IonObjectMapper/Serializer/IonObjectSerializer.cs

```diff
@@ -177,51 +195,101 @@ public override void Serialize(IIonWriter writer, object item)
             {
                 var ionFieldName = GetFieldName(field);
                 if (serializedIonFields.Contains(ionFieldName))
                 {
                     // This Ion field name was already serialized.
                     continue;
                 }
-                
-                if (options.IgnoreReadOnlyFields && field.IsInitOnly)
+
+                if (this.options.IgnoreReadOnlyFields && field.IsInitOnly)
                 {
                     continue;
                 }
 
                 var fieldValue = field.GetValue(item);
-                if (options.IgnoreNulls && fieldValue == null)
+                if (this.options.IgnoreNulls && fieldValue == null)
                 {
                     continue;
                 }
-                if (options.IgnoreDefaults && fieldValue == default)
+
+                if (this.options.IgnoreDefaults && fieldValue == default)
                 {
                     continue;
                 }
 
                 writer.SetFieldName(ionFieldName);
-                ionSerializer.Serialize(writer, fieldValue);
+                this.ionSerializer.Serialize(writer, fieldValue);
             }
 
             writer.StepOut();
         }
 
+        private static bool IsIonConstructor(ConstructorInfo constructor)
+        {
+            return constructor.GetCustomAttribute(typeof(IonConstructorAttribute)) != null;
+        }
+
+        private static bool IsIonField(FieldInfo field)
+        {
+            return field.GetCustomAttribute(typeof(IonFieldAttribute)) != null;
+        }
+
+        private static bool IsIonNamedProperty(PropertyInfo property)
+        {
+            return property.GetCustomAttribute(typeof(IonPropertyNameAttribute)) != null;
+        }
+
+        private static bool IsIonIgnore(PropertyInfo property)
+        {
+            return property.GetCustomAttribute(typeof(IonIgnoreAttribute)) != null;
+        }
+
+        /// <summary>
+        /// We only serialize public, internal, and protected internal properties or properties with IonPropertyName annotation.
+        /// </summary>
+        /// <param name="isGetter">Specify if we are looking for getters or setters in the property.</param>
+        private static bool HasValidAccessModifier(PropertyInfo propertyInfo, bool isGetter)
+        {
+            var methodInfo = isGetter ? propertyInfo.GetGetMethod(true) : propertyInfo.GetSetMethod(true);
+
+            return methodInfo != null && (methodInfo.IsPublic || methodInfo.IsAssembly || methodInfo.IsFamilyOrAssembly
+                || propertyInfo.GetCustomAttribute(typeof(IonPropertyNameAttribute)) != null);
+        }
+
+        private static bool IsReadOnlyProperty(PropertyInfo property)
+        {
+            return property.SetMethod == null;
+        }
+
+        private static string GetFieldName(FieldInfo field)
+        {
+            var propertyName = field.GetCustomAttribute(typeof(IonPropertyNameAttribute));
```

Comment:
If we did the `(IonPropertyNameAttribute)` cast on this line, then line 287 could be cleaned up to just `return propertyName.Name;`.

Reply:
Done.